### PR TITLE
frr: 9.1 -> 10.0

### PR DESCRIPTION
**Pull request by thillux**
## Description of changes

Release notes: https://github.com/FRRouting/frr/releases/tag/frr-10.0

Breaking changes:

- per-daemon config files no longer supported -> only when writing, not possible with configs in read-only store paths
- noprefixroute flag for interface prefixes with NetworkManager -> routes with noprefixroute set no longer inserted into advertised set, should be no real issue
- -n is required for mgmtd -> currently netns-based VRFs are not supported in the module. I'd ignore this until someone needs it explicitely.
- Enable enforce-first-as by default for BGP -> likely breaks Route Reflector/Route Server Setups -> mention in release notes?
- Deprecate ConfD -> confd is deprecated by its developing company tail-f

Currently WIP, opened for possible discussion.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by woffs** [comment]
The "breaking changes" have to be understood. https://github.com/NixOS/nixpkgs/pull/274425 has to be reviewed/merged.

**Comment by thillux** [comment]
@woffs What's your opinion regarding the breaking changes list?

**Comment by risicle** [comment]
Probably need some movement on this because of

CVE-2024-31948
CVE-2024-31949
CVE-2024-31950
CVE-2024-31951

**Comment by risicle** [comment]
Actually I notice the respective PRs for two of those CVEs haven't even been merged yet..

**Comment by woffs** [comment]
> @woffs What's your opinion regarding the breaking changes list?

per-daemon conf has to be replaced by unified config. this has to be reworked in the module, but seems to simplify things.

**Comment by thillux** [comment]
> > @woffs What's your opinion regarding the breaking changes list?
> 
> per-daemon conf has to be replaced by unified config. this has to be reworked in the module, but seems to simplify things.

FRR 10.0 also works fine with per-daemon config files in the NixOS setting. It is just not able to write them back to disk after changes in vtysh (which is nevertheless not possible on NixOS due to read-only config out of the store). I'd not change anything in the module for 10.0. Unifying the config to a single frr.conf will at least break user configs with the same route-map of prefix list names in different per-daemon config.

Let's justs state multiple config files as deprecated and unify them in the future otherwise it IMHO breaks config without necessity.

**Comment by thillux** [comment]
> Seems OK to me, but needs a release note if `enforce-first-as` is now the default for BGP. The note can then refer to the upstream changelog for other breaking changes, stating they do not impact users of the NixOS module (?)

Done

**Review by nicoonoclaste — APPROVED** [review]

**Comment by thillux** [comment]
> @thillux What's the plan to address the security issues in 23.11?

The plan is to address the security issues in 23.11 and unstable/24.05 when they have landed upstream. See e.g. https://github.com/FRRouting/frr/pull/15674 which is an unmerged fix for [CVE-2024-31951](https://nvd.nist.gov/vuln/detail/CVE-2024-31951)

File: pkgs/servers/frr/default.nix [hunk 1]
```diff
@@ -86,24 +86,15 @@ lib.warnIf (!(stdenv.buildPlatform.canExecute stdenv.hostPlatform))
...
     owner = "FRRouting";
     repo = pname;
     rev = "${pname}-${version}";
-    hash = "sha256-oDPr51vI+tlT1IiUPufmZh/UE0TNKWrn4RqpnGoGxNo=";
+    hash = "sha256-vvh9z2hmjvAA7OXgrUmlcrrTE5MRedZzfmhX5FEDKwE=";
   };
 
-  patches = [
-    # fixes crash in OSPF TE parsing
-    (fetchpatch {
```
Comment (SuperSandro2000) [inline pkgs/servers/frr/default.nix:100]: Not a pressing issue but it was forgotten to remove fetchpatch from the inputs.

**Pull request merged**